### PR TITLE
Introduce lowering_arena to avoid creating AST nodes on the fly

### PR DESCRIPTION
**Pull request by spastorino**
@oli-obk requested this and other changes as a way of simplifying https://github.com/rust-lang/rust/pull/101345. This is just going to make the diff of https://github.com/rust-lang/rust/pull/101345 smaller.

r? @oli-obk @cjgillot

### CODE REVIEW

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 1]
```diff
@@ -140,6 +148,12 @@ struct LoweringContext<'a, 'hir> {
     generics_def_id_map: Vec<FxHashMap<LocalDefId, LocalDefId>>,
 }
 
+struct LoweringArena {
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:151]: Could you use `rustc_arena::declare_arena`?

**Review by cjgillot — COMMENTED** [review]
Just one nit.  r=me afterwards.

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 2]
```diff
@@ -95,6 +96,13 @@ struct LoweringContext<'a, 'hir> {
     /// Used to allocate HIR nodes.
     arena: &'hir hir::Arena<'hir>,
 
+    /// Used to allocate temporary AST nodes for use during lowering.
+    /// This allows us to create "fake" AST -- these nodes can sometimes
+    /// be allocated on the stack, but other times we need them to live longer
+    /// than the current stack frame, so they can be collected into vectors
+    /// and things like that.
+    lowering_arena: &'a LoweringArena,
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:104]: Naming: everything is related to lowering here. `ast_arena`?

File: compiler/rustc_ast_lowering/src/lib.rs (hunk 1 above)
Comment (spastorino) [reply compiler/rustc_ast_lowering/src/lib.rs]: Done

File: compiler/rustc_ast_lowering/src/lib.rs (hunk 2 above)
Comment (spastorino) [reply compiler/rustc_ast_lowering/src/lib.rs]: Done

File: compiler/rustc_ast_lowering/src/lib.rs (hunk 1 above)
Comment (cjgillot) [reply compiler/rustc_ast_lowering/src/lib.rs]: I meant using the macro `rustc_arena::declare_arena`.

**Comment by spastorino** [comment]
> Just one nit. r=me afterwards.

@bors r=cjgillot

**Comment by spastorino** [comment]
@bors r-

I've figured that I misunderstood @cjgillot comment, gonna address the problem in a bit

File: compiler/rustc_ast_lowering/src/lib.rs (hunk 1 above)
Comment (spastorino) [reply compiler/rustc_ast_lowering/src/lib.rs]: I've tried to use `declare_arena` but that imposes a `'tcx` lifetime which any of the fields have. At that point I've tried `_marker: PhantomData<&'tcx ()>` but then I got a bunch of lifetime errors and I ended thinking that using the macro may be more complicated than writing this manually.
Let me know if you still want to persue this or this is ready to be merged in the way it is.

**Comment by spastorino** [comment]
I think this should be merged as is, commented on the problem with `declare_arena` in https://github.com/rust-lang/rust/pull/101499#discussion_r965282529

**Comment by spastorino** [comment]
@oli-obk said "leave a comment on the PhanomData field and r=me"

@bors r=oli-obk rollup

**Pull request merged**